lib/ansible/modules/files/copy.py

```diff
@@ -587,14 +587,16 @@ def main():
                 e.result['msg'] += ' Could not copy to {0}'.format(dest)
                 module.fail_json(**e.results)
 
-            os.makedirs(b_dirname)
+            if not module.check_mode:
```

Comment:
I haven't followed all the logic, but I'm concerned that this change will result in the `if os.path.isdir(b_dest)` conditional below behaving incorrectly under check mode, since the directory is no longer created:

https://github.com/ansible/ansible/blob/4e8b240b8b74db2453851f12ec82c7f782289187/lib/ansible/modules/files/copy.py#L599-L604